diffcp/cone_program.py

```diff
@@ -604,26 +621,140 @@ def adjoint_derivative(dx, dy, ds, **kwargs):
             perturbations; the sparsity pattern of `dA` matches that of `A`.
         """
 
-        dw = -(x @ dx + y @ dy + s @ ds)
-        dz = np.concatenate(
-            [dx, D_proj_dual_cone.rmatvec(dy + ds) - ds, np.array([dw])])
-
-        if np.allclose(dz, 0):
-            r = np.zeros(dz.shape)
-        elif mode == "dense":
-            r = _diffcp._solve_adjoint_derivative_dense(M, MT, dz)
-        elif mode == "lsqr":
-            r = _diffcp.lsqr(MT, dz).solution
-        elif mode == "lsmr":
-            MT_sp = sparse.linalg.LinearOperator(dz.shape*2, matvec=MT.matvec, rmatvec=MT.rmatvec)
-            r, istop, itn, normr, normar, norma, conda, normx = sparse.linalg.lsmr(MT_sp, dz, maxiter=10*MT_sp.shape[0], atol=1e-10, btol=1e-10)
-
-        values = pi_z[cols] * r[rows + n] - pi_z[n + rows] * r[cols]
-        dA = sparse.csc_matrix((values, (rows, cols)), shape=A.shape)
-        db = pi_z[n:n + m] * r[-1] - pi_z[-1] * r[n:n + m]
-        dc = pi_z[:n] * r[-1] - pi_z[-1] * r[:n]
+        if "lpgd" in mode:
+            dA, db, dc = adjoint_derivative_lpgd(dx, dy, ds, **derivative_kwargs, **kwargs)
+        else:
+            dw = -(x @ dx + y @ dy + s @ ds)
+            dz = np.concatenate(
+                [dx, D_proj_dual_cone.rmatvec(dy + ds) - ds, np.array([dw])])
+
+            if np.allclose(dz, 0):
+                r = np.zeros(dz.shape)
+            elif mode == "dense":
+                r = _diffcp._solve_adjoint_derivative_dense(M, MT, dz)
+            elif mode == "lsqr":
+                r = _diffcp.lsqr(MT, dz).solution
+            elif mode == "lsmr":
+                MT_sp = sparse.linalg.LinearOperator(dz.shape*2, matvec=MT.matvec, rmatvec=MT.rmatvec)
+                r, istop, itn, normr, normar, norma, conda, normx = sparse.linalg.lsmr(MT_sp, dz, maxiter=10*MT_sp.shape[0], atol=1e-10, btol=1e-10)
+
+            values = pi_z[cols] * r[rows + n] - pi_z[n + rows] * r[cols]
+            dA = sparse.csc_matrix((values, (rows, cols)), shape=A.shape)
+            db = pi_z[n:n + m] * r[-1] - pi_z[-1] * r[n:n + m]
+            dc = pi_z[:n] * r[-1] - pi_z[-1] * r[:n]
 
         return dA, db, dc
+    
+    def derivative_lpgd(dA, db, dc, tau, rho):
```

Comment:
Do you have quadratic objective support? I noticed you weren't banning them in construction above.

Reply:
Yes this is probably good to discuss. The LPGD method supports derivatives and adjoint derivatives for all parameters, including the quadratic P term here. The reason I didn't include it was to not break some compatibility by changing the arguments/outputs of the exposed derivative and adjoint_derivative methods (and their batched versions). I guess the derivative would be simple to resolve by just adding an optional dP=None argument, but for the adjoint derivative one would need to change the number of outputs. I was thinking about adding a request_dP flag? If you think this makes sense I can come up with some proposal changes.

Reply:
I added the differentiation w.r.t. P now, see the latest commits. I added two examples but have not tested this extensively